File: src/frontend/src/types/quest.ts
export type QuestType = 'daily' | 'weekly' | 'epic';

export interface QuestStep {
  description: string;
  condition_type: string;
  condition_value: number;
  completed: boolean;
}

export interface Quest {
  id: string;
  workspace_id: string;
  name: string;
  description: string;
  type: QuestType;
  steps: QuestStep[];
  xp_reward: number;
  currency_reward: number;
  icon: string;
  active: boolean;
  reset_hours: number | null;
  created_at: string;
}

export interface AgentQuestProgress {
  id: string;
  agent_id: string;
  quest_id: string;
  current_step: number;
  completed: boolean;
  completed_at: string | null;
  last_reset_at: string | null;
  quest: Quest;
}

File: src/backend/app/schemas/quest.py
"""Pydantic schemas for the quest progression system."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel


class QuestRead(BaseModel):
    id: str
    workspace_id: str
    name: str
    description: str
    type: str
    steps: Optional[Any] = None
    xp_reward: int
    currency_reward: int
    icon: str
    active: bool
    reset_hours: Optional[int] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class AgentQuestProgressRead(BaseModel):
    id: str
    agent_id: str
    quest_id: str
    current_step: int
    completed: bool
    claimed: bool
    completed_at: Optional[datetime] = None
    last_reset_at: Optional[datetime] = None
    quest: Optional[QuestRead] = None

    model_config = {"from_attributes": True}


class QuestWithProgressRead(QuestRead):
    """Quest including the requesting agent's progress, if any."""

    progress: Optional[AgentQuestProgressRead] = None
